### PR TITLE
Segment Replication - Fix NoSuchFileException errors caused when computing metadata snapshot on primary shards.

### PR DESCRIPTION
**Pull request by mch2**
Signed-off-by: Marc Handalian <handalm@amazon.com>

### Description
This change fixes errors that occur when computing metadata snapshots on primary shards from the latest in-memory SegmentInfos.  The error occurs when a segments_N file that is referenced by the in-memory infos is deleted as part of a commit.  The segments themselves are incref'd by IndexWriter.incRefDeleter but the commit file (Segments_N) is not.  This change resolves this by refactoring segment replication to ignore the segments_N file when computing metadata for CopyState and only sending incref'd segment files to replicas.

Given we are no longer sending the segments_N file, this change also updates the segmentReplicationDiff method to operate on Maps of SegmentMetadata instead of the MetadataSnapshot class.

As called out [here](https://github.com/opensearch-project/OpenSearch/issues/4178#issuecomment-1233820394) this chanage should be followed by an update on replicas to perform commits of the received SegmentInfos for durability purposes.

note - I've also added an assert on the index setting inside segrep specific store methods, requiring many test updates to set and pass the setting. 

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/4178

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/test/java/org/opensearch/index/store/StoreTests.java [hunk 1]
```diff
@@ -1150,9 +1158,52 @@ public void testGetMetadataWithSegmentInfos() throws IOException {
...
+        );
+        commitRandomDocs(store);
+
+        Store.MetadataSnapshot commitMetadata = store.getMetadata();
+
+        store.cleanupAndPreserveLatestCommitPoint("test", store.readLastCommittedSegmentsInfo());
+
+        // we want to ensure commitMetadata files are preserved after calling cleanup
+        for (String existingFile : store.directory().listAll()) {
+            assert (commitMetadata.contains(existingFile) == true);
```
[the remark this answers is not in the document]
Comment (mch2) [reply server/src/test/java/org/opensearch/index/store/StoreTests.java]: Thanks for calling this out , this test was never generating extra files and preserving the latest commit.  Updated so we flush some additional segments before this method is invoked.

**Comment by mch2** [comment]
@dreamer-89  Thanks for the review on this, I've made a couple of updates since you first looked, particularly on computing the segmentReplicationDiff from the Maps of StoreFileMetadata instead snapshots.  Am going to take this out of draft, I have put up another draft [here](https://github.com/opensearch-project/OpenSearch/pull/4402/files) that will update replicas to commit SegmentInfos when new generations are received from primaries.

**Comment by mch2** [comment]
```
  2> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.action.admin.indices.create.CreateIndexIT.testCreateAndDeleteIndexConcurrently" -Dtests.seed=34A84672B0BEDC12 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-CU -Dtests.timezone=Asia/Tashkent -Druntime.java=17
```

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 2]
```diff
@@ -334,6 +336,51 @@ public MetadataSnapshot getMetadata(SegmentInfos segmentInfos) throws IOExceptio
...
+    /**
+     * Segment Replication method
+     * Returns a diff between the Maps of StoreFileMetadata that can be used for getting list of files to copy over to a replica for segment replication. The returned diff will hold a list of files that are:
+     * <ul>
+     * <li>identical: they exist in both maps and they can be considered the same ie. they don't need to be recovered</li>
+     * <li>different: they exist in both maps but their they are not identical</li>
+     * <li>missing: files that exist in the source but not in the target</li>
+     * </ul>
+     */
+    public static RecoveryDiff segmentReplicationDiff(Map<String, StoreFileMetadata> source, Map<String, StoreFileMetadata> target) {
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/index/store/Store.java:358]: nit: Based on usage, only missing & different files are needed for segrep, is it better to modify the return type of this method ?

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 3]
```diff
@@ -825,17 +875,9 @@ public void commitSegmentInfos(SegmentInfos latestSegmentInfos, long maxSeqNo, l
...
-            assert existingCommits.isEmpty() == false : "Expected at least one commit but none found";
-            final IndexCommit lastIndexCommit = existingCommits.get(existingCommits.size() - 1);
-            assert latestSegmentInfos.getSegmentsFileName().equals(lastIndexCommit.getSegmentsFileName());
-            try (IndexWriter writer = newAppendingIndexWriter(directory, lastIndexCommit)) {
-                writer.setLiveCommitData(lastIndexCommit.getUserData().entrySet());
-                writer.commit();
-            }
+            directory.sync(latestSegmentInfos.files(true));
+            directory.syncMetaData();
+            cleanupAndPreserveLatestCommitPoint("After commit", latestSegmentInfos);
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/index/store/Store.java:880]: @mch2 : Why this change is needed ?

**Review by dreamer-89 — APPROVED** [review]
Thanks @mch2 for this change.

File: server/src/main/java/org/opensearch/index/store/Store.java (hunk 2 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/index/store/Store.java:358]: yes something we can consider as another refactor.

File: server/src/main/java/org/opensearch/index/store/Store.java (hunk 3 above)
Comment (mch2) [reply server/src/main/java/org/opensearch/index/store/Store.java:880]: the existing cleanup method was making another commit only for the fsync & deletion of old files. Any time this method was invoked it was moving gen + 2.  Here we fsync manually and are reusing our cleanup method that deletes unused files instead of making 2 "commits" in the same method.

File: server/src/main/java/org/opensearch/index/store/Store.java (hunk 2 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/index/store/Store.java:358]: I am fine taking this up in separate PR as this is not urgent.

File: server/src/main/java/org/opensearch/index/store/Store.java (hunk 3 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/index/store/Store.java:880]: Got it, thanks @mch2 for the clarification.
